File: src/services/diveLocationService.ts
/**
 * Dive Location Service
 * CRUD operations for dive locations in Firestore
 */

import { collection, doc, getDocs, getDoc, addDoc, updateDoc, deleteDoc, Timestamp, query, orderBy } from 'firebase/firestore';
import { db } from '@/lib/firebase';
import { DiveLocation } from '@/types/tariff.types';

export class DiveLocationService {
  /**
   * Get all locations for a club
   */
  static async getAllLocations(clubId: string): Promise<DiveLocation[]> {
    const locationsRef = collection(db, `clubs/${clubId}/dive_locations`);
    const q = query(locationsRef, orderBy('name', 'asc'));
    const snapshot = await getDocs(q);

    return snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data(),
      created_at: doc.data().created_at?.toDate() || new Date(),
      updated_at: doc.data().updated_at?.toDate() || new Date()
    })) as DiveLocation[];
  }

  /**
   * Get single location by ID
   */
  static async getLocationById(clubId: string, locationId: string): Promise<DiveLocation | null> {
    const docRef = doc(db, `clubs/${clubId}/dive_locations/${locationId}`);
    const docSnap = await getDoc(docRef);

    if (!docSnap.exists()) return null;

    return {
      id: docSnap.id,
      ...docSnap.data(),
      created_at: docSnap.data().created_at?.toDate() || new Date(),
      updated_at: docSnap.data().updated_at?.toDate() || new Date()
    } as DiveLocation;
  }

  /**
   * Create new location
   */
  static async createLocation(
    clubId: string,
    userId: string,
    data: Omit<DiveLocation, 'id' | 'created_at' | 'updated_at' | 'created_by'>
  ): Promise<string> {
    const locationsRef = collection(db, `clubs/${clubId}/dive_locations`);

    const docRef = await addDoc(locationsRef, {
      ...data,
      created_at: Timestamp.now(),
      updated_at: Timestamp.now(),
      created_by: userId
    });

    return docRef.id;
  }

  /**
   * Update existing location
   */
  static async updateLocation(
    clubId: string,
    locationId: string,
    updates: Partial<Omit<DiveLocation, 'id' | 'created_at' | 'updated_at' | 'created_by'>>
  ): Promise<void> {
    const docRef = doc(db, `clubs/${clubId}/dive_locations/${locationId}`);

    await updateDoc(docRef, {
      ...updates,
      updated_at: Timestamp.now()
    });
  }

  /**
   * Delete location
   */
  static async deleteLocation(clubId: string, locationId: string): Promise<void> {
    const docRef = doc(db, `clubs/${clubId}/dive_locations/${locationId}`);
    await deleteDoc(docRef);
  }
}

File: lib/services/dive_location_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import '../models/dive_location.dart';
import 'dive_location_sync_contract.dart';

/// Service pour les lieux de plongée
/// Firestore collection: clubs/{clubId}/dive_locations
class DiveLocationService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  /// Récupérer tous les lieux de plongée (ordonnés par nom)
  Future<List<DiveLocation>> getAllLocations(String clubId) async {
    final result = await getAllLocationsWithState(clubId);
    return result.items;
  }

  /// Charge le catalogue en distinguant données fraîches, cache, hors-ligne,
  /// erreur et catalogue réellement vide.
  Future<DiveLocationLoadResult<DiveLocation>> getAllLocationsWithState(
    String clubId,
  ) async {
    try {
      final snapshot = await _firestore
          .collection('clubs/$clubId/dive_locations')
          .orderBy('name', descending: false)
          .get();

      final locations = snapshot.docs
          .map((doc) => DiveLocation.fromFirestore(doc))
          .where((location) => location.mergedIntoLocationId == null)
          .toList();

      debugPrint('📍 ${locations.length} lieux de plongée chargés');
      return DiveLocationLoadResult(
        items: locations,
        state: stateForSnapshot(snapshot, locations.length),
      );
    } catch (error) {
      debugPrint('❌ Erreur chargement lieux: $error');
      try {
        final cachedSnapshot = await _firestore
            .collection('clubs/$clubId/dive_locations')
            .orderBy('name', descending: false)
            .get(const GetOptions(source: Source.cache));
        final cached = cachedSnapshot.docs
            .map((doc) => DiveLocation.fromFirestore(doc))
            .where((location) => location.mergedIntoLocationId == null)
            .toList();
        return DiveLocationLoadResult(
          items: cached,
          state: isTransientFirestoreError(error)
              ? DiveLocationLoadState.offline
              : (cached.isEmpty
                    ? DiveLocationLoadState.empty
                    : DiveLocationLoadState.cached),
          error: error,
        );
      } catch (_) {
        return DiveLocationLoadResult(
          items: const [],
          state: isTransientFirestoreError(error)
              ? DiveLocationLoadState.offline
              : DiveLocationLoadState.error,
          error: error,
        );
      }
    }
  }

  /// Récupérer uniquement les lieux proposés dans
  /// CalyCompta > Paramètres > Sites de plongée > Pour les activités.
  Future<List<DiveLocation>> getEventLocations(String clubId) async {
    final result = await getEventLocationsWithState(clubId);
    return result.items;
  }

  Future<DiveLocationLoadResult<DiveLocation>> getEventLocationsWithState(
    String clubId,
  ) async {
    final result = await getAllLocationsWithState(clubId);
    return DiveLocationLoadResult(
      items: result.items
          .where((location) => location.availableForEvents)
          .toList(),
      state: result.state,
      error: result.error,
    );
  }

  /// Récupérer un lieu par ID
  Future<DiveLocation?> getLocationById(
    String clubId,
    String locationId,
  ) async {
    var currentId = locationId;
    final visited = <String>{};
    for (var hop = 0; hop < 10 && !visited.contains(currentId); hop++) {
      visited.add(currentId);
      try {
        final doc = await _firestore
            .collection('clubs/$clubId/dive_locations')
            .doc(currentId)
            .get();
        if (!doc.exists) return null;
        final location = DiveLocation.fromFirestore(doc);
        final redirect = location.mergedIntoLocationId?.trim();
        if (redirect != null && redirect.isNotEmpty) {
          currentId = redirect;
          continue;
        }
        return location;
      } catch (error) {
        debugPrint('❌ Erreur chargement lieu: $error');
        return null;
      }
    }
    return null;
  }
}
